File: 39. Combination Sum.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) :
        
        start = 0 
        z = []
        rest = target
        ans = []
        n = len(nums)
        def dfs(start,z,rest):
            if rest == 0 :
                ans.append(z[:])
                return
            
            for i in range(start,n):
                x =  nums[i]
                val =  rest - x 
                if val >=0 :
                    dfs(i,z+[x],val)
            
        dfs(0,z,rest)
        return ans

File: 39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int> candidates, int target) {       sort(candidates.begin(),candidates.end());
        vector<vector<int>> ans;
        vector<int>temp;
        helper(target,candidates,ans,temp,0);
        return ans;
        
    }
    void helper(int target,vector<int> candidates,vector<vector<int>> &ans,vector<int> temp,int start){
         if (target == 0){
            ans.push_back(temp);
            return;
        }
        if (target <  candidates[start]) {
            return ;
        }
       
       for(int i = start ; i <candidates.size();i++){
              temp.push_back(candidates[i]);
              helper(target-candidates[i],candidates,ans,temp,i);  
              temp.pop_back();
            }      
    }
    
};
